src/plugins/wizard/public/embeddable/wizard_embeddable.tsx

```diff
@@ -3,125 +3,288 @@
+    const ExpressionLoader = getExpressionLoader();
+    this.handler = new ExpressionLoader(this.node, undefined, {
+      onRenderError: (_element: HTMLElement, error: ExpressionRenderError) => {
+        this.onContainerError(error);
+      },
+    });
+
+    if (this.savedWizard?.description) {
+      div.setAttribute('data-description', this.savedWizard.description);
+    }
+
+    div.setAttribute('data-test-subj', 'wizardLoader');
+
+    this.subscriptions.push(this.handler.loading$.subscribe(this.onContainerLoading));
+    this.subscriptions.push(this.handler.render$.subscribe(this.onContainerRender));
+
+    this.updateHandler();
   }
 
-  /**
-   * Lets re-sync our saved object to make sure it's up to date!
-   */
   public async reload() {
-    this.savedObjectId = this.input.savedObjectId;
-    const wizardSavedObject = await this.savedObjectsClient.get<WizardSavedObjectAttributes>(
-      'wizard',
-      this.input.savedObjectId
-    );
-    const savedAttributes = wizardSavedObject?.attributes;
-    this.updateOutput({
-      savedAttributes,
-      title: wizardSavedObject?.attributes?.title,
-    });
+    this.updateHandler();
   }
 
   public destroy() {
     super.destroy();
-    this.subscription.unsubscribe();
+    this.subscriptions.forEach((s) => s.unsubscribe());
     if (this.node) {
       ReactDOM.unmountComponentAtNode(this.node);
     }
+
+    if (this.handler) {
+      this.handler.destroy();
+      this.handler.getElement().remove();
+    }
+    this.autoRefreshFetchSubscription.unsubscribe();
+  }
+
+  private async updateHandler() {
+    const expressionParams: IExpressionLoaderParams = {
+      searchContext: {
+        timeRange: this.timeRange,
+        query: this.input.query,
+        filters: this.input.filters,
+      },
+    };
+    if (this.abortController) {
+      this.abortController.abort();
+    }
+    this.abortController = new AbortController();
+    const abortController = this.abortController;
+
+    if (this.handler && !abortController.signal.aborted) {
+      this.handler.update(this.expression, expressionParams);
+    }
+  }
+
+  public async handleChanges() {
+    // TODO: refactor (here and in visualize) to remove lodash dependency - immer probably a better choice
+
+    let dirty = false;
+
+    // Check if timerange has changed
+    if (!_.isEqual(this.input.timeRange, this.timeRange)) {
+      this.timeRange = _.cloneDeep(this.input.timeRange);
+      dirty = true;
+    }
+
+    // Check if filters has changed
+    if (!opensearchFilters.onlyDisabledFiltersChanged(this.input.filters, this.filters)) {
+      this.filters = this.input.filters;
+      dirty = true;
+    }
+
+    // Check if query has changed
+    if (!_.isEqual(this.input.query, this.query)) {
+      this.query = this.input.query;
+      dirty = true;
+    }
+
+    // Check if rootState has changed
+    if (!_.isEqual(this.getSerializedState(), this.serializedState)) {
+      this.serializedState = this.getSerializedState();
```

Comment:
If `getSerializedState` isn't extremely cheap, I would suggest calling it only once, storing the result, and then comparing and assigning.

Reply:
It is very cheap - just popping some props with fallbacks (see L105-112). In fact, it's use here is because it's the cheapest thing to check for changes.